### PR TITLE
feat: add published_time publisher debug to packages

### PR DESCRIPTION
**Pull request by brkay54**
## Description

Part of:
- https://github.com/autowarefoundation/autoware.universe/issues/6255

Depends on:
- https://github.com/autowarefoundation/autoware.universe/pull/6440 

## Tests performed
Tested on PSim and e2e_simulator, worked well.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by mitsudome-r** [comment]
@miursh @TakaHoribe I would like to have your comments as perception/control/planning modules as a whole rather than asking confirmation from each package owners. Are you okay with the design?

**Comment by TakaHoribe** [comment]
@mitsudome-r As an overall design, it looks ok to me.

**Comment by mitsudome-r** [comment]
Please check README of each node and make sure that published topics are also updated.

**Comment by xmfcx** [comment]
- https://results.pre-commit.ci/run/github/429651616/1710332674.AOFM05LoQ4CPpeg5aPvnYw

failed, pushed this commit to fix it: https://github.com/autowarefoundation/autoware.universe/pull/6490/commits/49e7c45e48e57ef911a6928cb9a3d8c8f26d16be

**Comment by xmfcx** [comment]
- https://github.com/autowarefoundation/autoware.universe/actions/runs/8264643291/job/22608668912?pr=6490#step:8:2488

This failed but probably a flaky test. Opened an issue for it:
- https://github.com/autowarefoundation/autoware.universe/issues/6612

File: perception/detected_object_feature_remover/README.md [hunk 1]
```diff
@@ -22,6 +22,4 @@ The `detected_object_feature_remover` is a package to convert topic-type from `D
 
 ## Parameters
 
-None
```
Comment (xmfcx) [inline perception/detected_object_feature_remover/README.md:25]: Is this necessary?

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp [hunk 2]
```diff
@@ -199,6 +200,12 @@ class MapBasedPredictionNode : public rclcpp::Node
   std::vector<double> distance_set_for_no_intention_to_walk_;
   std::vector<double> timeout_set_for_no_intention_to_walk_;
 
+  // Stop watch
+  StopWatch<std::chrono::milliseconds> stop_watch_;
+
```
Comment (xmfcx) [inline perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp:205]:
```suggestion
```

Unused variable.

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp [hunk 3]
```diff
@@ -199,6 +200,12 @@ class MapBasedPredictionNode : public rclcpp::Node
...
+  // PublishedTime Publisher
```
Comment (xmfcx) [inline perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp:206]:
```suggestion
```

These comments are not useful, they are stating the obvious. Could you remove all instances of these comments?

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 4]
```diff
@@ -1112,6 +1114,10 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
 
   // Publish Results
   pub_objects_->publish(output);
+
+  // Publish published time only if there are subscribers more than 1
+  published_time_publisher_->publish(pub_objects_, output.header.stamp);
```
Comment (xmfcx) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:1119]: We should have named this method as `publish_if_subscribed()`

Could you refactor this function name?
No need to open a new PR, you can do it within this PR.

File: perception/tracking_object_merger/src/decorative_tracker_merger.cpp [hunk 5]
```diff
@@ -220,8 +222,12 @@ void DecorativeTrackerMergerNode::mainObjectsCallback(
 
   // try to merge main object
   this->decorativeMerger(main_sensor_type_, main_objects);
+  const auto & tracked_objects = getTrackedObjects(main_objects->header);
+  merged_object_pub_->publish(tracked_objects);
```
Comment (xmfcx) [inline perception/tracking_object_merger/src/decorative_tracker_merger.cpp:226]:
```suggestion
```

Why are you publishing these again?

Comment (xmfcx) [reply perception/tracking_object_merger/src/decorative_tracker_merger.cpp:226]: If you need a reference to the tracked_objects, you can keep this but you should remove the L228 then.

File: planning/path_smoother/README.md [hunk 6]
```diff
@@ -9,3 +9,5 @@ This package contains code to smooth a path or trajectory.
 ### Elastic Band
 
 More details about the elastic band can be found [here](docs/eb.md).
+
+## Node parameters
```
Comment (xmfcx) [inline planning/path_smoother/README.md:13]:
```suggestion
```

Unnecessary addition?

File: control/trajectory_follower_node/src/controller_node.cpp [hunk 7]
```diff
@@ -231,7 +233,10 @@ void Controller::callbackTimerControl()
   out.longitudinal = lon_out.control_cmd;
   control_cmd_pub_->publish(out);
 
-  // 6. publish debug marker
+  // 6. publish published time only if there are subscribers more than 1
+  published_time_publisher_->publish(control_cmd_pub_, out.stamp);
+
+  // 7. publish debug marker
```
Comment (xmfcx) [inline control/trajectory_follower_node/src/controller_node.cpp:239]:
```suggestion
```
We should remove unnecessary comment lines.

Comment (brkay54) [reply control/trajectory_follower_node/src/controller_node.cpp:239]: Sorry, I couldn't understand what is unnecessary here. The comments are ordered in code, I just added another step here:

```
  // 6. publish published time only if there are subscribers more than 1
  published_time_publisher_->publish(control_cmd_pub_, out.stamp);

  // 7. publish debug marker
  publishDebugMarker(*input_data, lat_out);
```

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 4 above)
Comment (brkay54) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: Hi, I agree with you. I changed the function name. 

Also PR for adding unit test was updated: ```publish() -> publish_if_subscribed()```
- https://github.com/autowarefoundation/autoware.universe/pull/6610

File: control/trajectory_follower_node/src/controller_node.cpp [hunk 8]
```diff
@@ -231,7 +233,8 @@ void Controller::callbackTimerControl()
   out.longitudinal = lon_out.control_cmd;
   control_cmd_pub_->publish(out);
 
-  // 6. publish debug marker
+  // 6. publish debug
```
Comment (xmfcx) [inline control/trajectory_follower_node/src/controller_node.cpp:236]:
```suggestion
```

this comment line is unnecessary

Comment (brkay54) [reply control/trajectory_follower_node/src/controller_node.cpp:236]: I couldn't understand here. The code owner was ordered the steps via comments so I did not want to delete this line.

**Original code:**
```
  // 4. sync with each other controllers
  longitudinal_controller_->sync(lat_out.sync_data);
  lateral_controller_->sync(lon_out.sync_data);

  // TODO(Horibe): Think specification. This comes from the old implementation.
  if (isTimeOut(lon_out, lat_out)) return;

  // 5. publish control command
  autoware_auto_control_msgs::msg::AckermannControlCommand out;
  out.stamp = this->now();
  out.lateral = lat_out.control_cmd;
  out.longitudinal = lon_out.control_cmd;
  control_cmd_pub_->publish(out);

  // 6. publish debug marker
  publishDebugMarker(*input_data, lat_out);
  
  ```

Comment (xmfcx) [reply control/trajectory_follower_node/src/controller_node.cpp:236]: fine 🙂‍↔️

**Comment by xmfcx** [comment]
@brkay54 Could you go through all the changed packages to make sure they have ` <depend>tier4_autoware_utils</depend>` in their package.xml? (just to make sure)

**Comment by brkay54** [comment]
@xmfcx Thank you for the detailed review. 🙏🏼 

> @brkay54 Could you go through all the changed packages to make sure they have ` <depend>tier4_autoware_utils</depend>` in their package.xml? (just to make sure)

Just checked the missed ones, I added them with the last commit.

**Comment by xmfcx** [comment]
@brkay54 thanks,

Right now only way for this PR to break existing behavior or even crash the modified nodes is:

If `published_time_publisher_` is not initialized in a constructor and if it is called from somewhere, it will cause segfault.

Could you make sure every modified node initializes this `published_time_publisher_` object?

I don't know if there is a way to check this in a more automated way.

**Comment by xmfcx** [comment]
I've checked all modified `.cpp` files manually to make sure they are all initialized in the constructors, it should be ok to merge this now.

Waiting for the re-run test that had failed from an unrelated flaky test.

**Pull request merged**